### DISCUSSION TITLE
How to Execute a ScriptBlock with SessionState correctly inside a ICommandPredictor?

### DISCUSSION THREAD

**Discussion by AnderssonPeter** [PowerShell/PowerShell#16527] [Q&A] [answered]
Hi I'm trying to execute a `ScriptBlock` with a specific `SessionState` inside a `ICommandPredictor`, but if the `ScriptBlock` has a pipe in it just hangs forever (Could be a deadlock).

I have a project that recreates the issue [here](https://github.com/AnderssonPeter/PowerShellScriptBlockHang).

To recreate the issue:
1. Clone the repo
2. Open in Visual Studio 2021
3. Set a breakpoint in `Predictor.cs` on [line 23](https://github.com/AnderssonPeter/PowerShellScriptBlockHang/blob/1c5476681695216c14fdf12dd528084ac422aae7/PowerShellScriptBlockHang/Predictor.cs#L23)
4. Start with debugging
5. Type one char into the terminal that opens up.
6. Visual studio should popup with the breakpoint hit
7. Press F10 (Step over)
8. You will never reach [line 28](https://github.com/AnderssonPeter/PowerShellScriptBlockHang/blob/1c5476681695216c14fdf12dd528084ac422aae7/PowerShellScriptBlockHang/Predictor.cs#L28) or [31](https://github.com/AnderssonPeter/PowerShellScriptBlockHang/blob/1c5476681695216c14fdf12dd528084ac422aae7/PowerShellScriptBlockHang/Predictor.cs#L31) as we are stuck.

So I might have done something wrong in how I execute ScriptBlocks..
But the flow is:
1. When the Terminal starts it executes [Install-ScriptBlock](https://github.com/AnderssonPeter/PowerShellScriptBlockHang/blob/main/PowerShellScriptBlockHang/InstallScriptBlock.cs)
    1. This exectues a `ScriptBlock` with a new Session that in turn executes [set-scriptblock](https://github.com/AnderssonPeter/PowerShellScriptBlockHang/blob/main/PowerShellScriptBlockHang/SetScriptBlock.cs), `set-scriptblock` stores the provided `ScriptBlock` and the new `SessionState`.
    2. Create a ICommandPredictor and register it with the Subsystem.
2. When you type in the terminal the `ICommandPredictor` is invoked, this in turn tries to execute the scriptblock that was saved with `set-scriptblock` using the same session.

I hope this makes sense and I know the flow is somewhat convoluted, but I can't remove any of the steps in the flow without breaking my use case..

So is there any way I could change this to allow me to execute a `ScriptBlock` inside a `ICommandPredictor` with a `SessionState` that was created when executing a `ScriptBlock` without causing it to hang?

**Comment by SeeminglyScience** [comment]
See [this thread](https://github.com/PowerShell/PowerShell/pull/13186#discussion_r456148007).

TL;DR: The predictor API was not designed to be dependent on runspace state due to time sensitivity concerns. You would have to strip `SessionState` affinity and invoke in a new runspace.

**Reply by AnderssonPeter** [reply]
How would i go about stripping it of the affinity? Sadly I don't know that much of PowerShell internals :(

**Reply by AnderssonPeter** [reply]
Just thinking aloud here would it be a better solution to create my own runspace in `InstallScriptBlock.cs` instead of using `InvokeCommand.InvokeScript`?

**Reply by SeeminglyScience** [reply]
Easiest way is to just `ToString()` it and pass it in `PowerShell.AddScript`. You will definitely need to create a separate runspace for it. Note that in general you may have a tough time writing PowerShell code of even medium complexity that is able to consistently keep up with the 20ms timeout.

**Comment by AnderssonPeter** [comment, marked as answer]
After talking to SeeminglyScience on discord he explained why it is a bad idea to execute this ondemand in a `ICommandPredictor`!
